### PR TITLE
[GEOS-10598] XSS vulnerability in the email address field

### PR DESCRIPTION
**Pull request by ahmedababnehgeo**
[GEOS-10598](https://osgeo-org.atlassian.net/browse/GEOS-10598)

In this PR, we preform output encoding to protect against XSS attacks. Furthermore, we do validation on the email field to reduce the risk of attacks. Several techniques, to protect against these attacks can be found here: https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html 

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by ianturton — APPROVED** [review]

**Comment by ahmedababnehgeo** [comment]
@aaime PR is ready for review

File: src/main/src/main/java/org/geoserver/config/impl/ContactInfoImpl.java [hunk 1]
```diff
@@ -152,7 +154,12 @@ public void setAddressType(String addressType) {
 
     @Override
     public String getContactEmail() {
-        return InternationalStringUtils.getOrDefault(contactEmail, internationalContactEmail);
+        String result = contactEmail;
+        if (result == null && internationalContactEmail != null)
+            result = internationalContactEmail.toString(GeoServerDefaultLocale.get());
+        String escapedResult = StringEscapeUtils.escapeHtml4(result);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/config/impl/ContactInfoImpl.java:160]: This should be done in the UI and in the REST freemarker templates, but not here. For example, a REST call getting the email field, returned in XML or JSON, should not have escaping happening, the field should be returned as-is in those cases.

File: src/web/core/src/main/java/org/geoserver/web/StringAndInternationalStringPanel.java [hunk 2]
```diff
@@ -55,4 +56,38 @@ protected TextField<String> getTextComponent(String id, IModel<String> model) {
...
+    public StringAndInternationalStringPanel(
+            String id,
+            IModel<?> model,
+            String labelKey,
+            String stringProperty,
+            String internationalProperty,
+            WebMarkupContainer labelProvider,
+            IValidator<String> validator) {
+        super(id, model);
+        WebMarkupContainer container = new WebMarkupContainer("labelContainer");
```
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/StringAndInternationalStringPanel.java:69]: The addition of validators is done in the right places, but duplicating the entire existing constructor to add them is overkill. I suggest using a single constructor with actual code, a couple of ifs, and the previous constructor would call the new one passing null as the validator.

**Review by aaime — CHANGES_REQUESTED** [review]
Why have the two tests for WMS and WMTS store pages been removed?
I don't agree with the solution, should not be affecting how the mail is saved and handed in core, but only touching the sensitive HTML parts:
- https://github.com/geoserver/geoserver/blob/fe4a8909d51df12f13dc0f3c7e368780add2a6e2/src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java#L68
- Potentially the freemarker templates, if they are not correctly escaping already

File: src/web/core/src/main/java/org/geoserver/web/StringAndInternationalStringPanel.java (hunk 2 above)
Comment (ahmedababnehgeo) [reply src/web/core/src/main/java/org/geoserver/web/StringAndInternationalStringPanel.java:39]: I removed the extra constructor as suggested

**Comment by ahmedababnehgeo** [comment]
@aaime PR has been updated.

File: src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java [hunk 3]
```diff
@@ -76,7 +76,7 @@ public GeoServerHomePage() {
                             "footerMessage",
                             new StringResourceModel(
                                     "GeoServerHomePage.footer", this, new Model(params)));
-            label.setEscapeModelStrings(false);
+            label.setEscapeModelStrings(true);
```
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java:79]: Yup, thanks

File: src/wfs/src/main/java/org/geoserver/wfs/CapabilitiesTransformer.java [hunk 4]
```diff
@@ -1392,8 +1395,9 @@ protected void internationalServiceProvider(ContactInfo contact) {
                                 contact.getInternationalAddressCountry()));
                 element(
                         "ows:ElectronicMailAddress",
-                        internationalContentHelper.getNullableString(
-                                contact.getInternationalContactEmail()));
+                        StringEscapeUtils.escapeXml10(
```
Comment (aaime) [inline src/wfs/src/main/java/org/geoserver/wfs/CapabilitiesTransformer.java:1398]: The contact mail is used in other capabilities document generators as well (WCS, WMS) but wondering if this escaping is actually useful, I think the low-level XML encoder here should already be doing the necessary escaping for you. Do you have a way to actually do harm using the XML documents?

Comment (ahmedababnehgeo) [reply src/wfs/src/main/java/org/geoserver/wfs/CapabilitiesTransformer.java:1398]: Good point ... I am unaware of a way to attack using XML document? Do you suggest rolling back this change?!

Comment (aaime) [reply src/wfs/src/main/java/org/geoserver/wfs/CapabilitiesTransformer.java:1398]: If it's not needed, better to back it out, yes (if you find a way to move an attack using XML documents instead, it needs tests and be replicated in all other places a caps document emits the mail address).

**Comment by ahmedababnehgeo** [comment]
@aaime PR ready for review

**Pull request merged**